src/librustc_expand/mbe/macro_rules.rs

```diff
@@ -184,12 +184,41 @@ fn generic_extension<'cx>(
 ) -> Box<dyn MacResult + 'cx> {
     if cx.trace_macros() {
         let msg = format!("expanding `{}! {{ {} }}`", name, pprust::tts_to_string(arg.clone()));
-        trace_macros_note(cx, sp, msg);
+        trace_macros_note(&mut cx.expansions, sp, msg);
     }
 
     // Which arm's failure should we report? (the one furthest along)
     let mut best_failure: Option<(Token, &str)> = None;
+
+    // We create a base parser that can be used for the "black box" parts.
+    // Every iteration needs a fresh copy of that base parser. However, the
+    // parser is not mutated on many of the iterations, particularly when
+    // dealing with macros like this:
+    //
+    // macro_rules! foo {
+    //     ("a") => (A);
+    //     ("b") => (B);
+    //     ("c") => (C);
+    //     // ... etc. (maybe hundreds more)
+    // }
+    //
+    // as seen in the `html5ever` benchmark. We use a `Cow` so that the base
+    // parser is only cloned when necessary (upon mutation). Furthermore, when
+    // that happens, we discard the mutated parser and reinitialize the `Cow`
+    // with the base parser before the next iteration. This is all quite hacky,
+    // but speeds up the `html5ever` benchmark significantly. (Issue 68836
+    // suggests a more comprehensive but more complex change to deal with this
+    // situation.)
+    let base_parser = base_parser_from_cx(&cx.current_expansion, &cx.parse_sess, arg.clone());
+    let mut parser = Cow::Borrowed(&base_parser);
+
     for (i, lhs) in lhses.iter().enumerate() {
+        if parser.is_owned() {
```

Comment:
The condition is unnecessary, the assignment inside it will be a no-op if it returns `false`.

Reply:
True. I did as you suggested, and this let me move `parser` within the loop, which also removed the need for the `drop(parser)`.

Reply:
I worried that the assignment will generate a memcpy copying padding for the whole `Cow<Parser>` which is huge because `Parser` is huge.
Looks like that's not the case though, only the useful payload (~2 pointers) is copied.